### DISCUSSION TITLE
How to build a button that adds bookings to a booking overview page

### DISCUSSION THREAD

**Discussion by liebeckm** [neuefische/questions#566] [Web] [answered]
Hello again,

I am building a carpool app. Therefore, I want to have a page within my app that shows all rides that have been booked by the user already. I already have an [overview page ](https://github.com/liebeckm/capstone-project/blob/booking-confirmation/pages/index.js)of all offered rides, a [detail page ](https://github.com/liebeckm/capstone-project/blob/booking-confirmation/pages/detail-ride/%5Bslug%5D.js)of each offered ride and a [booking confirmation page](https://github.com/liebeckm/capstone-project/blob/booking-confirmation/pages/detail-ride/booking-confirmation/%5Bslug%5D.js).

I would like to implement a code that shows the booked ride on a ["my-rides"-page](https://github.com/liebeckm/capstone-project/blob/booking-confirmation/pages/my-rides/index.js) as soon as the user hit the "book"-link.

For that, I already did the following:
1. I tried to work with useState on the details page, where the booking link appears
2. 1. I made a new folder "my-rides" and a index.js within it, in there, I tried to map through all booked rides. 

when I type http://localhost:3000/my-rides, the following error message occurs:
TypeError: bookedRides.map is not a function

I would like to know how to solve this problem; Also, I am unsure about the structure, where to write which code and last but not least, in the detail-ride page, I wrote a const "handleConfirmBooking" but I don't know where to use it. Maybe it's not necessary at all?

Can someone of you guys help me out here? Thank you so much!
If needed, [this is the user story for this](https://github.com/liebeckm/capstone-project/issues/3)

**Comment by lenzloesche** [comment]
hi,
you need to uplift the bookedRides to the _app.js so you can use it on all pages. The _app.js is the page that gets loaded, whenever you load any page.

in your _app.js:
`import {useState} from "react";`
and
`const [bookedRides, setBookedRides] = useState([]);`
and 
`bookedRides={bookedRides} setBookedRides={setBookedRides}`

then import it on your my-rides-page like so:
`export default function BookedRidesPage({bookedRides}) {`

also this before the other return statement so there is no error with the empty bookedRides: 
```js
if (!bookedRides) {
    return;
  }

  if (bookedRides.length === 0) {
    return;
  }
  ````
in your slug.js you import the bookedRides:
`export default function DetailPage({ride, bookedRides, setBookedRides}) {`

and change your handleBookRide to this:
```js
const handleBookRide = (event) => {
    event.preventDefault(); //this is so you dont get to reload the page.
    setBookedRides([...bookedRides, extractedRide]);
    router.push("/my-rides");  //this is only to change the site so you can test the page
  };
  ````
and change this:
´onClick={(event) => { handleBookRide(event);}}´

you might want to consider using local storage state and global state.

hope it helps.

**Reply by liebeckm** [reply]
Hey @lenzloesche thank you for the detailed explanation! I think we are on the right way here but anyways, it's not working right now.

1. I can't write `export default function DetailPage({ ride, bookedRides, setBookedRides })` on the details page; It throws two comments; ride is declared but it's value is never read and: "Identifier "bookedRides" has already been declared." Also the browser throws an error message:

`./pages/detail-ride/[slug].js
Error: 
  x cannot reassign to a variable declared with `const`
    ,-[/Users/melissa/Documents/neuefische/projects/capstone-project/pages/detail-ride/[slug].js:20:1]
 20 |   StyledLinkWrapper,
 21 | } from "../../components/Link/Link.styled";
 22 | 
 23 | export default function DetailPage({ ride, bookedRides, setBookedRides }) {
    :                                            ^^^^^|^^^^^
    :                                                 `-- cannot reassign
 24 |   const router = useRouter();
 25 |   const { slug } = router.query;
 26 |   const extractedRide = rides.find((ride) => ride.slug === slug);
 27 |   const [bookedRides, setBookedRides] = useState([]);
    :          ^^^^^|^^^^^
    :               `-- const variable was declared here
 28 | 
 29 |   const handleBookRide = (event) => {
 30 |     event.preventDefault();
    `----

  x cannot reassign to a variable declared with `const`
    ,-[/Users/melissa/Documents/neuefische/projects/capstone-project/pages/detail-ride/[slug].js:20:1]
 20 |   StyledLinkWrapper,
 21 | } from "../../components/Link/Link.styled";
 22 | 
 23 | export default function DetailPage({ ride, bookedRides, setBookedRides }) {
    :                                                         ^^^^^^^|^^^^^^
    :                                                                `-- cannot reassign
 24 |   const router = useRouter();
 25 |   const { slug } = router.query;
 26 |   const extractedRide = rides.find((ride) => ride.slug === slug);
 27 |   const [bookedRides, setBookedRides] = useState([]);
    :                       ^^^^^^^|^^^^^^
    :                              `-- const variable was declared here
 28 | 
 29 |   const handleBookRide = (event) => {
 30 |     event.preventDefault();
    `----

  x the name `bookedRides` is defined multiple times
    ,-[/Users/melissa/Documents/neuefische/projects/capstone-project/pages/detail-ride/[slug].js:20:1]
 20 |   StyledLinkWrapper,
 21 | } from "../../components/Link/Link.styled";
 22 | 
 23 | export default function DetailPage({ ride, bookedRides, setBookedRides }) {
    :                                            ^^^^^|^^^^^
    :                                                 `-- previous definition of `bookedRides` here
 24 |   const router = useRouter();
 25 |   const { slug } = router.query;
 26 |   const extractedRide = rides.find((ride) => ride.slug === slug);
 27 |   const [bookedRides, setBookedRides] = useState([]);
    :          ^^^^^|^^^^^
    :               `-- `bookedRides` redefined here
 28 | 
 29 |   const handleBookRide = (event) => {
 30 |     event.preventDefault();
    `----

  x the name `setBookedRides` is defined multiple times
    ,-[/Users/melissa/Documents/neuefische/projects/capstone-project/pages/detail-ride/[slug].js:20:1]
 20 |   StyledLinkWrapper,
 21 | } from "../../components/Link/Link.styled";
 22 | 
 23 | export default function DetailPage({ ride, bookedRides, setBookedRides }) {
    :                                                         ^^^^^^^|^^^^^^
    :                                                                `-- previous definition of `setBookedRides` here
 24 |   const router = useRouter();
 25 |   const { slug } = router.query;
 26 |   const extractedRide = rides.find((ride) => ride.slug === slug);
 27 |   const [bookedRides, setBookedRides] = useState([]);
    :                       ^^^^^^^|^^^^^^
    :                              `-- `setBookedRides` redefined here
 28 | 
 29 |   const handleBookRide = (event) => {
 30 |     event.preventDefault();
    `----`

2. If I delete the `{ ride, bookedRides, setBookedRides }` on the details page, I tried the following:  if I use this code 
`const handleBookRide = (event) => {
    event.preventDefault(); //this is so you don't get to reload the page.
    setBookedRides([...bookedRides, extractedRide]);
    router.push("/my-rides");  //this is only to change the site so you can test the page
  };`
and
`     <StyledLink
            href={`./booking-confirmation/${slug}`}
            onClick={(event) => {
              handleBookRide(event);
            }}
          >
            Buchen
          </StyledLink>`
on my details page and click on the link, I will go directly to http://localhost:3000/my-rides ; but normally I'd like to go to the booking confirmation page. Anyways, when clicking on the link, the booked ride should appear on the my-rides page!

**Reply by lenzloesche** [reply]
ok, I forgot to tell you to delete the declaration of the state in the details page.
when uplifting the state, you have to delete this in the Detail page.
`const [bookedRides, setBookedRides] = useState([]);`
Because you now have the bookedRides variable in the _app.js and you send it as a prop down to the details page, so would have it as a prop AND as a new variable and that doesnt work.

the ride variable, i think you dont need it. so you only need:
`export default function DetailPage({ bookedRides, setBookedRides })`

EDIT: you can also look into the session handouts of react state. To understand it better.

**Reply by liebeckm** [reply]
Thank you @lenzloesche !! I tried to read all the react state handouts but find it hard to apply this on my specific case :( 

I tried two things now:
1.  On details page, I deleted `const [bookedRides, setBookedRides] = useState([]);`; on _app.js I wrote the following code: `import GlobalStyle from "../styles";
import Head from "next/head";
import { useState } from "react";

export default function App({ Component, pageProps }) {
  const [bookedRides = { bookedRides }, setBookedRides = { setBookedRides }] =
    useState([]);

  return ... 
` Error message is: 
`Unhandled Runtime Error
TypeError: bookedRides is not iterable

Source
pages/detail-ride/[slug].js (30:23) @ bookedRides

  28 | const handleBookRide = (event) => {
  29 |   event.preventDefault();
> 30 |   setBookedRides([...bookedRides, extractedRide]);
     |                     ^
  31 |   router.push("/my-rides");
  32 | };
  33 | ``

2. On details page, I deleted `const [bookedRides, setBookedRides] = useState([]);`;I wrote the bookedRides and setBookRides on _app.js as the following:
`export default function App({ Component, pageProps }) {
  const [bookedRides, setBookedRides] = useState([]);
  bookedRides={bookedRides};
  setBookedRides={setBookedRides}
...`
Error message is:
`./pages/_app.js
Error: 
  x cannot reassign to a variable declared with `const`
    ,-[/Users/melissa/Documents/neuefische/projects/capstone-project/pages/_app.js:3:1]
  3 | import { useState } from "react";
  4 | 
  5 | export default function App({ Component, pageProps }) {
  6 |   const [bookedRides, setBookedRides] = useState([]);
    :          ^^^^^|^^^^^
    :               `-- const variable was declared here
  7 |   bookedRides = { bookedRides };
    :   ^^^^^|^^^^^
    :        `-- cannot reassign
  8 |   setBookedRides = { setBookedRides };
  9 | 
 10 |   return (
    `----

  x cannot reassign to a variable declared with `const`
    ,-[/Users/melissa/Documents/neuefische/projects/capstone-project/pages/_app.js:3:1]
  3 | import { useState } from "react";
  4 | 
  5 | export default function App({ Component, pageProps }) {
  6 |   const [bookedRides, setBookedRides] = useState([]);
    :                       ^^^^^^^|^^^^^^
    :                              `-- const variable was declared here
  7 |   bookedRides = { bookedRides };
  8 |   setBookedRides = { setBookedRides };
    :   ^^^^^^^|^^^^^^
    :          `-- cannot reassign
  9 | 
 10 |   return (
 11 |     <>
    `----`

[I also pushed the code again, for better understanding.](https://github.com/liebeckm/capstone-project/tree/booking-confirmation)

**Reply by lenzloesche** [reply]
I looked at your code in the booking-confirmation branch.

I didnt make myself clear. you have to write it in the app.js like this:
```js
 <Component
        {...pageProps}
        bookedRides={bookedRides}
        setBookedRides={setBookedRides}
      />
```
**not** like this how you did it:
```js
  [...]
  bookedRides = { bookedRides };
  setBookedRides = { setBookedRides };
  return [...]
```

**Comment by liebeckm** [comment, marked as answer]
Hey @lenzloesche good point! :) Now it's working – I also had do delete the event in order to use the "buchen" button in the right way, which is:
click buchen button -> redirect to booking confirmation page -> in the same time, add the booked ride to the my rides page. I made a small "my rides" button on the main page to test this. Now I will build a navbar for that.

I'm going to mark this discussion as solved :) Thank you for your explanations!
